server/src/session_manager.ts

```diff
@@ -16,70 +26,209 @@ export interface YoutubeSessionDataCaches {
+        sourceAddress,
+        disableTlsVerification,
+    }: Partial<ProxySpec>) {
+        this.proxy = proxy;
+        this.sourceAddress = sourceAddress;
+        this.disableTlsVerification = disableTlsVerification || false;
     }
+    public asDispatcher(logger: Logger): Agent | undefined {
+        const { proxy, sourceAddress, disableTlsVerification } = this;
+        let sanitizedProxy = proxy;
+        if (!sanitizedProxy) {
+            return new Agent({
+                localAddress: sourceAddress,
+                rejectUnauthorized: !disableTlsVerification,
+            });
+        }
+        let protocol: string;
+        try {
+            const parsedUrl = new URL(sanitizedProxy);
+            protocol = parsedUrl.protocol.replace(":", "");
+            // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        } catch (e) {
+            // assume http if no protocol was passed
+            protocol = "http";
+            sanitizedProxy = `http://${sanitizedProxy}`;
+        }
 
-    log(msg: string) {
-        if (this.shouldLog) console.log(msg);
-    }
+        let loggedProxy: string = sanitizedProxy;
+        try {
+            const parsedUrl = new URL(sanitizedProxy);
+            if (parsedUrl.password) {
+                loggedProxy = sanitizedProxy.replace(
+                    parsedUrl.password,
+                    "****",
+                );
+            }
+        } catch (e) {
+            logger.warn(`Fail to parse proxy url ${sanitizedProxy}: ${e}`);
+            return undefined;
+        }
 
-    warn(msg: string) {
-        // stderr should always be shown
-        console.warn(msg);
+        switch (protocol) {
+            case "http":
+            case "https":
+                logger.log(`Using HTTP/HTTPS proxy: ${loggedProxy}`);
+                return new HttpsProxyAgent(sanitizedProxy, {
+                    rejectUnauthorized: !disableTlsVerification,
+                    localAddress: sourceAddress,
+                });
+            case "socks":
+            case "socks4":
+            case "socks4a":
+            case "socks5":
+            case "socks5h": {
+                logger.log(`Using SOCKS proxy: ${loggedProxy}`);
+                const agent = new SocksProxyAgent(sanitizedProxy);
+                agent.options.localAddress = sourceAddress;
+                agent.options.rejectUnauthorized = !disableTlsVerification;
+                return agent;
+            }
+            default:
+                logger.warn(`Unsupported proxy protocol: ${loggedProxy}`);
+                return undefined;
+        }
     }
+}
 
-    error(msg: string) {
-        console.error(msg);
+class CacheSpec {
+    constructor(
+        public pxySpec: ProxySpec,
+        public ip: string | null,
+    ) {}
+    public get key(): string {
+        return JSON.stringify(
+            this.ip || [this.pxySpec.proxy, this.pxySpec.sourceAddress],
+        );
     }
 }
 
+type CachedTokenMinter = {
```

Comment:
Isn't this just a TokenMinter? Why do we need to called it cached

Reply:
In my opinion, things aren't "cached" until they're stored in a cache, with the intent of being retrieved later. Similar to how a PO token has an inherent expiry, but we don't really consider it 'cached' until we store it in our cache with the intent of retrieving it later